### PR TITLE
CMS: Add testable combobox component

### PR DESCRIPTION
**Pull request by DirkDevelopmentFreelance**

### CODE REVIEW

**Review by ivanslabbert — CHANGES_REQUESTED** [review]
Blocking merging until reviewed by Johan and checked by Ivan

File: cms/src/components/forms/LCombobox.vue [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    showGroups.value = !showGroups.value;
+    input.value.focus();
+};
+
+onClickOutside(groupsDisplay, () => (showGroups.value = false));
+</script>
+
+<template>
+    <div class="relative" :class="$attrs['class']" :style="$attrs['style'] as StyleValue">
+        <FormLabel> Group Membership </FormLabel>
```
Comment (belljohan3) [inline cms/src/components/forms/LCombobox.vue:77]:
```suggestion
        <FormLabel> Group membership </FormLabel>
```

**Review by belljohan3 — CHANGES_REQUESTED** [review]
It works well but we need to do more on UX.

- When clicking on a group in the list, the dropdown list is still showing
- When searching for a group in the input, the dropdown is in full height. 
image

**Review by belljohan3 — APPROVED** [review]
Nice job

**Review by ivanslabbert — CHANGES_REQUESTED** [review]
Hi Dirk

This looks good, but...
The Combobox is now specific for showing groups and not a generic component that we can use for any combobox implementation. We will now have to copy this code and modify it for tag selectors and any other selectors we need.

Would it be possible to make it a generic combobox (a bit like HeadlessUI have done) that we can use for different combobox use cases?

File: cms/src/components/content/EditContentParent.vue [hunk 2]
```diff
@@ -15,6 +15,7 @@ import { capitaliseFirstLetter } from "@/util/string";
 import FormLabel from "@/components/forms/FormLabel.vue";
 import LToggle from "@/components/forms/LToggle.vue";
 import ImageEditor from "../images/ImageEditor.vue";
+import LCombobox from "../forms/LCombobox.vue";
```
Comment (ivanslabbert) [inline cms/src/components/content/EditContentParent.vue:18]: Linting error

File: cms/src/components/forms/LCombobox.spec.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+            await wrapper.find("li").trigger("click");
+
+            const lTag = wrapper.findComponent(LTag);
+
+            await waitForExpect(() => {
+                expect(lTag.html()).toContain("Test Label");
+            });
+        },
+        { timeout: 100000 },
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.spec.ts:61]: This is a very long timeout? Guess it was used during debugging?

File: cms/src/components/forms/LCombobox.vue [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    label?: string;
+    disabled?: boolean;
+    options: ComboboxOption[];
+};
+
+const props = withDefaults(defineProps<Props>(), {
+    disabled: false,
+});
+
+//const selected = ref<Array<string | number>>([])
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:26]: remove commented code?

File: cms/src/components/forms/LCombobox.vue [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            o.label.toLowerCase().includes(query.value.toLowerCase()),
+        );
+    } else {
+        optionsToDisplay.value = currentOptionsList;
+    }
+});
+
+const input = ref();
+
+const optionsDisplay = ref();
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:58]: The naming is a bit confusing (as we also have an `optionsToDisplay`). Perhaps something like `optionsElement` or `optionsDiv` could work?

File: cms/src/components/forms/LCombobox.vue [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                        {
+                            'text-zinc-500 hover:bg-white': selectedOptions?.includes(option.id),
+                        },
+                    ]"
+                    @click="
+                        () => {
+                            if (!selectedOptions?.includes(option.id)) {
+                                console.log('ADDED');
+                                selectedOptions?.push(option.id);
+                                console.log(selectedOptions);
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:118]: remove console.log

File: cms/src/components/forms/LCombobox.vue [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+const optionsList = computed(() =>
+    props.options.map((o) => ({
+        id: o.id,
+        label: o.label,
+        value: o.value,
+        selected: selectedOptions.value?.includes(o.id),
+    })),
+);
+const optionsToDisplay = ref(optionsList.value);
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:38]: Not sure why we need this? A computed is already a ref.

File: cms/src/components/forms/LCombobox.vue [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            ref="optionsDisplay"
+            v-show="showOptions"
+            class="absolute z-10 mt-1 max-h-48 w-full overflow-y-auto rounded-md border-[1px] border-zinc-100 bg-white shadow-md"
+            data-test="options"
+        >
+            <ul>
+                <li
+                    v-for="option in optionsToDisplay"
+                    :key="option.id"
+                    :disabled="selectedOptions?.includes(option.id)"
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:100]: You should already have a `selected` parameter in `optionsToDisplay` which removes the need to recalculate it here?
```suggestion
                    :disabled="option.selected"
```

File: cms/src/components/forms/LCombobox.vue [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                    class="text-sm hover:bg-zinc-100"
+                    :class="[
+                        'relative cursor-default select-none py-2 pl-3 pr-9',
+                        {
+                            'bg-white text-black hover:bg-zinc-100': !selectedOptions?.includes(
+                                option.id,
+                            ),
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:107]:
```suggestion
                            'bg-white text-black hover:bg-zinc-100': !option.selected,
```

File: cms/src/components/forms/LCombobox.vue [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                        },
+                        {
+                            'text-zinc-500 hover:bg-white': selectedOptions?.includes(option.id),
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:110]:
```suggestion
                            'text-zinc-500 hover:bg-white': option.selected,
```

File: cms/src/components/forms/LCombobox.vue [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                        },
+                    ]"
+                    @click="
+                        () => {
+                            if (!selectedOptions?.includes(option.id)) {
```
Comment (ivanslabbert) [inline cms/src/components/forms/LCombobox.vue:115]:
```suggestion
                            if (!option.selected) {
```

File: cms/src/components/languages/CreateOrEditLanguageModal.vue [hunk 12]
```diff
@@ -91,6 +92,26 @@ const validateForm = () => {
...
+        label: "Untranslated",
+    },
+    {
+        id: 2,
+        value: "all",
+        label: "All",
+    },
+];
+
+const selected = ref([]);
```
Comment (ivanslabbert) [inline cms/src/components/languages/CreateOrEditLanguageModal.vue:114]: Guess you would still do some cleanup here...

**Review by ivanslabbert — APPROVED** [review]

**Pull request merged**